Video promos 66

evanbret opened the pull request:
## Description
_Information about what you changed for this PR_
1. Added headline position custom field to medium promo, small promo, and top table list. (Headline can be before or after image.)
2. Upped potential number of small images per row in top table list to 4.
3. Implemented Video promo item in large and extra large blocks in top table list.

## Jira Ticket
- [SENG-66](https://arcpublishing.atlassian.net/browse/SENG-66)

## Acceptance Criteria
1. PageBuilder users can configure Small and Medium Promo Blocks, as well as Top Table List, to display the promo image above or below the headline
2. PageBuilder users can configure Small items in the Top Table List to display up to 4 stories per row
(Video implementation was outside scope of ticket but was asked to do this.)

## Test Steps
- Add test steps a reviewer must complete to test this PR
Add medium promo, small promo, and top table list to page. Give content: for brands site, medium and small promos can have 
UO2TYMABUJDCFODJJQ4FB57BOA as the story ID and the top table list can select section feed with the section /main. Play around with configurations: whether image is above or below headline, and how many small per row in the top table list.

## Dependencies or Side Effects

(Images currently broken, think this may be an issue with sdk Image component.)

## Review Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [ ] Confirmed all the test steps above are working
- [ ] Confirmed there are no linter errors
- [ ] Confirmed this PR has reasonable code coverage
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [ ] Confirmed relevant documentation has been updated/added.

evanbret committed:
added position options to promo blocks

evanbret committed:
added video sdk component to top table

evanbret committed:
Merge branch 'canary' into video-promos-66

evanbret committed:
fixed some errors

LukeMason commented:
@evanbret can you correct linting errors, the formatting changes make this PR a lot harder to read and also break linting. `npm run lint:fix` should fix many but not all errors automatically.

JackHowa commented on blocks/default-output-block/package-lock.json:
```diff
@@ -13,11 +13,11 @@
-      "version": "7.11.0",
-      "resolved": "https://registry.npmjs.org/@babel/generator/-/generator-7.11.0.tgz",
-      "integrity": "sha512-fEm3Uzw7Mc9Xi//qU20cBKatTfs2aOtKqmvy/Vm7RkJEGFQ4xc9myCfbXxqK//ZS8MR/ciOHw6meGASJuKmDfQ==",
+      "version": "7.11.6",
```
You should not need to update package dependencies for this ticket. Can you revert the changes on the package-lock. In actuality (for better and for worse), package-lock files aren't published to blocks

JackHowa commented:
@evanbret, these are the known preexisting warnings in canary. there should be no new ones: 

```
$ npm run lint

> fusion-news-theme-blocks@ lint /Users/howardj/sites/fusion-news-theme-blocks
> eslint --ext js --ext jsx blocks components --no-error-on-unmatched-pattern


/Users/howardj/sites/fusion-news-theme-blocks/blocks/header-nav-block/features/navigation/_children/search-box.jsx
  17:23  warning  Assignments to the 'disabledBtn' variable from inside React Hook useEffect will be lost after each render. To preserve the value over time, store it in a useRef Hook and keep the mutable value in the '.current' property. Otherwise, you can move this variable directly inside useEffect  react-hooks/exhaustive-deps

/Users/howardj/sites/fusion-news-theme-blocks/blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/search-box.jsx
  17:23  warning  Assignments to the 'disabledBtn' variable from inside React Hook useEffect will be lost after each render. To preserve the value over time, store it in a useRef Hook and keep the mutable value in the '.current' property. Otherwise, you can move this variable directly inside useEffect  react-hooks/exhaustive-deps

/Users/howardj/sites/fusion-news-theme-blocks/blocks/right-rail-advanced-block/layouts/right-rail-advanced/default.jsx
  8:27  warning  React Hook "useAppContext" is called in function "getFeatureList" that is neither a React function component nor a custom React Hook function  react-hooks/rules-of-hooks

/Users/howardj/sites/fusion-news-theme-blocks/blocks/right-rail-block/layouts/right-rail/default.jsx
  7:27  warning  React Hook "useAppContext" is called in function "getFeatureList" that is neither a React function component nor a custom React Hook function  react-hooks/rules-of-hooks

/Users/howardj/sites/fusion-news-theme-blocks/blocks/video-player-block/features/video-player/default.jsx
  27:53  warning  React Hook "useContent" is called conditionally. React Hooks must be called in the exact same order in every component render. Did you accidentally call a React Hook after an early return?  react-hooks/rules-of-hooks

/Users/howardj/sites/fusion-news-theme-blocks/blocks/video-promo-block/features/video-promo/default.jsx
  39:33  warning  React Hook "useContent" is called conditionally. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks

✖ 6 problems (0 errors, 6 warnings)

evanbret committed:
addressed linting errors

evanbret committed:
undoing package change

evanbret committed:
small fix

evanbret committed:
tiny tiny fix

JackHowa commented:
Looks like the tests are failing @evanbret. Did you see our pr checklist? sorry if it's tedious to go over the steps

JackHowa commented:
$ `npm t`

```
blocks/article-body-block/chains/article-body/_children/table.test.jsx

Summary of all failing tests
 FAIL  blocks/small-promo-block/features/small-promo/default.test.jsx
  ● the small promo feature › should have two link elements by default

    expect(received).toHaveLength(expected)

    Expected length: 2
    Received length: 1
    Received object: {}

      46 |   it('should have two link elements by default', () => {
      47 |     const wrapper = mount(<SmallPromo customFields={config} />);
    > 48 |     expect(wrapper.find('a')).toHaveLength(2);
         |                               ^
      49 |   });
      50 | 
      51 |   it('should link the headline to the current site website_url ANS property', () => {

      at Object.<anonymous> (features/small-promo/default.test.jsx:48:31)

  ● the small promo feature › should link the headline to the current site website_url ANS property

    toHaveProp must be called on a single node, not an empty node.

      52 |     const url = mockData.websites['the-sun'].website_url;
      53 |     const wrapper = mount(<SmallPromo customFields={config} />);
    > 54 |     expect(wrapper.find('a.sm-promo-headline')).toHaveProp('href', url);
         |                                                 ^
      55 |   });
      56 | 
      57 |   it('should link the image to the current site website_url ANS property', () => {

      at Object.<anonymous> (features/small-promo/default.test.jsx:54:49)

  ● the small promo feature › should link the image to the current site website_url ANS property

    toHaveProp must be called on a single node, not an empty node.

      58 |     const url = mockData.websites['the-sun'].website_url;
      59 |     const wrapper = mount(<SmallPromo customFields={config} />);
    > 60 |     expect(wrapper.find('a').at(1)).toHaveProp('href', url);
         |                                     ^
      61 |   });
      62 | 
      63 |   it('should have one img when show image is true', () => {

      at Object.<anonymous> (features/small-promo/default.test.jsx:60:37)

  ● the small promo feature › Headline div should have class .col-sm-xl-8 when show image is true

    expect(received).toHaveLength(expected)

    Expected length: 1
    Received length: 0
    Received object: {}

      68 |   it('Headline div should have class .col-sm-xl-8 when show image is true', () => {
      69 |     const wrapper = mount(<SmallPromo customFields={config} />);
    > 70 |     expect(wrapper.find('.col-sm-xl-8')).toHaveLength(1);
         |                                          ^
      71 |   });
      72 | 
      73 |   it('should have no Image when show image is false', () => {

      at Object.<anonymous> (features/small-promo/default.test.jsx:70:42)

  ● the small promo feature › headline div should have class .col-sm-xl-12 when show image is false

    expect(received).toHaveLength(expected)

    Expected length: 1
    Received length: 0
    Received object: {}

      88 |     };
      89 |     const wrapper = mount(<SmallPromo customFields={noImgConfig} />);
    > 90 |     expect(wrapper.find('.col-sm-xl-12')).toHaveLength(1);
         |                                           ^
      91 |   });
      92 | 
      93 |   it('should only be one link when showHeadline is false and show image is true', () => {

      at Object.<anonymous> (features/small-promo/default.test.jsx:90:43)

 FAIL  blocks/medium-promo-block/features/medium-promo/default.test.jsx
  ● the medium promo feature › should have two link elements by default

    expect(received).toHaveLength(expected)

    Expected length: 2
    Received length: 1
    Received object: {}

      46 |   it('should have two link elements by default', () => {
      47 |     const wrapper = mount(<MediumPromo customFields={config} />);
    > 48 |     expect(wrapper.find('a')).toHaveLength(2);
         |                               ^
      49 |   });
      50 | 
      51 |   it('should link the headline to the current site website_url ANS property', () => {

      at Object.<anonymous> (features/medium-promo/default.test.jsx:48:31)

  ● the medium promo feature › should link the headline to the current site website_url ANS property

    toHaveProp must be called on a single node, not an empty node.

      52 |     const url = mockData.websites['the-sun'].website_url;
      53 |     const wrapper = mount(<MediumPromo customFields={config} />);
    > 54 |     expect(wrapper.find('a.md-promo-headline')).toHaveProp('href', url);
         |                                                 ^
      55 |   });
      56 | 
      57 |   it('should link the image to the current site website_url ANS property', () => {

      at Object.<anonymous> (features/medium-promo/default.test.jsx:54:49)

  ● the medium promo feature › should link the image to the current site website_url ANS property

    toHaveProp must be called on a single node, not an empty node.

      58 |     const url = mockData.websites['the-sun'].website_url;
      59 |     const wrapper = mount(<MediumPromo customFields={config} />);
    > 60 |     expect(wrapper.find('a').at(1)).toHaveProp('href', url);
         |                                     ^
      61 |   });
      62 | 
      63 |   it('should have one img when show image is true', () => {

      at Object.<anonymous> (features/medium-promo/default.test.jsx:60:37)

 FAIL  blocks/top-table-list-block/features/top-table-list/_children/medium-list-item.test.jsx
  ● medium list item › renders title and image with full props

    Method “text” is meant to be run on 1 node. 0 found instead.

      80 | 
      81 |     // finds description text
    > 82 |     expect(wrapper.find('p.description-text').text()).toBe(descriptionText);
         |                                               ^
      83 | 
      84 |     expect(wrapper.find('MediumListItem > hr').length).toBe(1);
      85 |   });

      at ReactWrapper.single (../../node_modules/enzyme/src/ReactWrapper.js:1168:13)
      at ReactWrapper.text (../../node_modules/enzyme/src/ReactWrapper.js:629:17)
      at Object.<anonymous> (features/top-table-list/_children/medium-list-item.test.jsx:82:47)

 FAIL  blocks/top-table-list-block/features/top-table-list/_children/vertical-overline-image-story-item.test.jsx
  ● vertical overline image story item › renders image and overline with full props

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      90 | 
      91 |     // has the correct link
    > 92 |     expect(wrapper.find('a.xl-promo-headline').length).toBe(1);
         |                                                        ^
      93 |     expect(wrapper.props().websiteURL).toBe('url');
      94 |     expect(wrapper.find('a.xl-promo-headline').props().href).toBe(websiteURL);
      95 | 

      at Object.<anonymous> (features/top-table-list/_children/vertical-overline-image-story-item.test.jsx:92:56)


Test Suites: 4 failed, 96 passed, 100 total
Tests:       10 failed, 1 skipped, 829 passed, 840 total
Snapshots:   2 passed, 2 total
Time:        30.291s, estimated 37s
Ran all test suites in 56 projects.
npm ERR! Test failed.  See above for more details.
it-20750:fusion-news-theme-blocks howardj$

JackHowa commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -117,6 +130,20 @@ const MediumPromo = ({ customFields }) => {
     <>
       <article className="container-fluid medium-promo">
         <div className={`medium-promo-wrapper ${customFields.showImage ? 'md-promo-image' : ''}`}>
+          {customFields.headlinePosition === 'above'
```
Can you fix existing tests and add logic for using headlinePosition? if you're unsure about testing, react testing library has incredible docs. If you want as well, I'd be happy to screenshare. My slack is @jackhowa as well

JackHowa replied:
rtl docs https://testing-library.com/docs/react-testing-library/intro

JackHowa commented:
thanks for fixing the linting errors and reverting all package-locks. sorry, it makes it easier for reviewers to look-over quickly

evanbret committed:
test fixes

evanbret committed:
undoing changes

brentmiller1973 committed:
fixed lint errors in search-result.test.jsx

brentmiller1973 commented on blocks/medium-promo-block/features/medium-promo/default.jsx:
```diff
@@ -36,26 +36,39 @@ const MediumPromo = ({ customFields }) => {
+    && content.credits.by
+    && content.credits.by.length !== 0
+    ? content.credits.by
+    : null;
```
For lines 48-57 would suggest using optional chaining here to make it more clean:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

brentmiller1973 commented on blocks/top-table-list-block/features/top-table-list/_children/horizontal-overline-image-story-item.jsx:
```diff
@@ -100,65 +106,114 @@ const HorizontalOverlineImageStoryItem = (props) => {
+    && element.promo_items
+    && element.promo_items.basic
+    && element.promo_items.basic.additional_properties
+    && element.promo_items.basic.additional_properties.videoId;
```
Another opportunity to use optional chaining here:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

brentmiller1973 reviewed (changes requested):
See my 2 comments on using optional chaining.  I committed a lint fix to blocks/search-results-list-block/features/search-results-list/_children/search-result.test.jsx. Did you add the tests needed to cover headline position?  That needs to be covered.

evanbret committed:
implemented optional chaining

evanbret committed:
merging in lintfix change from Brent
Merge branch 'video-promos-66' of https://github.com/WPMedia/fusion-news-theme-blocks into video-promos-66

evanbret committed:
small adjustments

evanbret committed:
small adjustments again

evanbret committed:
Merge branch 'canary' into video-promos-66

JackHowa reviewed (commented):
thanks @evanbret -- sorry retracting my 'request changes' after tests addressed
